### PR TITLE
Added fleet jump functionality. Shift + Jump key engages fleet jump.

### PR DESCRIPTION
**Pull request by Jeremy-E-Johnson**
Hitting jump key again either jumps with ready ships or engages standard
ASAP style autopilot if flagship is not ready.

### CODE REVIEW

**Comment by tehhowch** [comment]
Why'd you close this PR? It's best to PR from a feature branch, not the branch you use to track `es/master`. If your content doesn't get merged, you'll have a lot more hoops to jump through to offer other contributions, plus you won't have extraneous commits in your PR that are simply you updating your `master` with the latest updates to `es/master`. The mergetool is smart about what gets added, and you don't have to keep updating your PR branch with `es/master` content unless you have merge conflicts.

**Comment by Jeremy-E-Johnson** [comment]
Ah that makes sense. Thanks for the suggestion! Sorry about that, I only have a little version control experience.

I'll close the other one now, and move my post describing it over to here:

This PR is directly related to #2638.

Under these changes hitting shift + jump button activates fleet jump autopilot. The fleet will then prepare to jump but will wait once ready. The player then can hit the jump button to jump with all ready ships. Hitting the jump button again before the flag ship is ready causes the auto pilot to revert to the standard ASAP jump.

The fleet jump activation is remembered in between jumps as long as the flagship was ready to jump when the jump key was hit.

It seems to play nicely with wormholes and autopilot landings.

The only issue that I know of is in the lack of explanation about fleet jump. Outside of the message that is displayed when it is engaged, there are no other references to it supplied to the player. I considered adding a new "with " box in the controls menu, but I have not done so. If this is desirable then I can update this pull request with such a change. I believe that it would either come in the form of expanding the current "with " box or by adding a new one.

File: source/AI.cpp [hunk 1]
```diff
@@ -2667,26 +2698,33 @@ void AI::MovePlayer(Ship &ship, const PlayerInfo &player)
...
 			keyStuck.Clear();
+			shiftStuck = false;
 			if(keyDown.Has(Command::JUMP) || !keyHeld.Has(Command::JUMP))
 				Audio::Play(Audio::Get("fail"));
 		}
 		else
 		{
 			PrepareForHyperspace(ship, command);
 			command |= Command::JUMP;
-			if(keyHeld.Has(Command::JUMP))
```
Comment (tehhowch) [inline source/AI.cpp]: Could this just be
`if(keyHeld.Has(Command::JUMP) || shiftStuck)`
e.g. "If you are holding down Jump, or if you have indicated fleet jump, then we should wait"

Comment (Jeremy-E-Johnson) [inline source/AI.cpp]: If we use that if statement then if shiftStuck is true we will not jump no matter what. Even if the flagship is ready and the player is pressing the jump key. We need to make sure that is shiftStuck is true and the jump key is being pressed that either the autopilot is canceled or the ready ships jump.

**Comment by Forlarren** [comment]
This almost solves the "just jump ahead and you are invincible" problem. Unless I'm wrong and it does, I have to wait for release to test.

Assuming I'm understanding this right, one request if it's possible. Can the shift-j combo only order freighters to get positioned?

That way escorts keep engaged while the slow vulnerable freighters get prepped.

Now if there was only a way to make non-combat ships take automatic damage when you leave them behind the skip ahead to be invincible problem is solved.

The big choice will be do you run as soon as your freighters are prepped and hope you don't run into more trouble in the next system or do you collect the fleet so escorts can rush the enemy before they get to the cargo haulers. A meaningful decision.

Could the jump key be interrupted until all freighters are ready? So the player is first in last out in regards to just escorted (freighter class) ships in a fleet, so you can't leave without your freighters. I don't even bother trading anymore I just edit my save for money because there isn't any point as cheating is a core mechanic, avoiding content is way easier than dealing with it. Makes it pointless to even try being a realistic freighter fleet commodore.

Maybe an alt-j option is the new behavior and shift-j is all ships. Just as long as you can't leave without anyone you are escorting (freighters and escorts if either is even possible). That would close the biggest gaping hole in the game right now (in my opinion at least).

**Comment by Jeremy-E-Johnson** [comment]
> This almost solves the "just jump ahead and you are invincible" problem. Unless I'm wrong and it does, I have to wait for release to test.

Unfortunately, this does not really address that problem. However, it does add some functionality that will be useful if that is fixed. All this pull request does is change the control scheme. It adds the ability to hit shift + jump and for that to act as though you are holding the jump key until you press the jump key again.

As for only ordering freighters to get positioned, that may be possible, but it would require more code changes and would be a more challenging change than what is currently here.

The fleet jump is interrupted in the same way as any other autopilot cancellation. For example, hitting the forward key would stop it and all ships in your fleet will stop getting ready for jump.

Some sort of selective alt-j option would be interesting. It would be less straight forward to program, but could be worth it. However, I don't know if I feel conferrable attempting to address the issue of abandoning escorts without reaching a clear community consensus on how that problem should be solved. (especially as I am new here)

**Comment by Forlarren** [comment]
> The fleet jump is interrupted in the same way as any other autopilot cancellation. For example, hitting the forward key would stop it and all ships in your fleet will stop getting ready for jump.

That is so annoying. It's killing all the fun for me. Thanks for fixing the hold J thing, everyone with a RSI will love it. And at least I can force myself to pretend running away without my fleet is impossible. Holding the down the "don't cheat" dead mans switch was too much.

I'm super new to git, can I get this without recompiling? On windows for now.

**Comment by Jeremy-E-Johnson** [comment]
As of right now, (and to my somewhat limited understanding) you would have to merge the branch from my repo into your local repo. Then you would have to recompile before the changes showed up at run-time. Though you could probably get away with just recompiling the AI.h and AI.cpp files as they were all that was changed.

**Comment by tehhowch** [comment]
@Forlarren You can add this to a branch on your fork by navigating to the Jeremy's fork, selecting the branch in question (`fleet-jump-develop`), and then clicking the `Pull Request` button on the right. You will need to change the "base fork" as it will default to this repository, and not your own. You should then change the branch on your fork to match the one you are making your changes with.
This can be done a little more clearly with Git Shell, as you would `cd` into your repo, create a remote pointing to Jeremy's fork (`git remote add .....`), switch to your desired branch (`git checkout MiniSky`), obtain the details of your remote (`git fetch ....`), and then `git pull <remote name>/fleet-jump-develop`

@Jeremy-E-Johnson those would need to be compiled, along with any file which includes "AI.h" as a dependency. After compiling, comes linking (making the binary from all the `*.o` files.

**Pull request closed without merging**